### DISCUSSION TITLE
V-cycle errors

### DISCUSSION THREAD

**Discussion by Julian-Wyatt** [wenceorg/comp4187#86] [Q&A] [answered]
I've managed to successfully pass the jacobi tests for the coursework (ie the test_two_grid function), however my v-cycle code produces increasing errors on more refined grids which look like:

[1.18167568e-05, 1.99099355e-06, 2.68594848e-07, 1.28529688e-01, 3.97952194e+04],
on grid refinements 1-5, when solved with LU decomposition at the coarsest level (running jacboi here instead just performs slightly worse).
When calculating the W-cycle, I get very similar errors.

I'm not sure what I could be missing here, would you be able to give any advice or insight?

**Comment by annereinarz** [comment, marked as answer]
There's a few things that could be causing issues/that you can try:
- increase the number of smoothing steps: does the convergence continue further?
- change the alpha scaling for the Jacobi iterations.
- can you check if the error is highest at the boundaries? Try scaling both sides of the equations by the 3D volume element.
- check the restriction/prolongation operators you are using. Injection as a restriction might not work particularly well.

**Reply by Julian-Wyatt** [reply]
Thanks - this is for the mms convergence test on the 1D Poisson solver with the preset parameters. I've changed the parameters as suggested - for results below

## 1.
Altering the pre and post smoothing steps from 1 to 10 and trying post smoothing = 2x pre smoothing, I found that the overall results were of the same magnitude (ie grid refinements 4 and 5 still perform poorly).
Increasing pre and post iters to 40 gives errors of: 
V cycle:
[1.73367637e-05 3.07660989e-06 4.58071118e-07 1.54047477e-07 8.96348653e-01]
W cycle:
[1.73212064e-05 3.13555370e-06 5.65832698e-07 1.79710076e-07 4.76815470e+01]
## 2.
Updated alpha to 0.94 (with pre and post smoothing iters kept at 1) gives better error of:
V cycle:
[1.72803171e-05 2.99676583e-06 4.58914367e-07 2.12145402e-05 2.63942233e+01]
W cycle:
[1.73402772e-05, 3.14048444e-06, 4.05319078e-07, 1.15997288e-08, 8.92409541e+00]


## 3.
For 4 degrees of freedom (ie 1d grid of size 161), and after 10 iterations of the loop in MGSolver.solve, the error (expectation - x) is 0.005 (at x_{10}), increasing linearly to 0.07 (at x_{140}), which then decreases to 0.015 at x_{157}, so it doesn't look like the majority of the error is at the boundary. As the iterations tends to 60+, the error generally stays of +- same order of magnitude, until about 70-80 where the error increases past 1.

**Reply by annereinarz** [reply]
It looks like you've found a better alpha value, that's good. I'd try to stick with somewhere in the range of 3-5 smoothing iterations. For 3 it doesn't look like that's the issue, but just multiplying both sides of the equation with the volume element might still help, so maybe give that a go. 
In the 1D test case there is a factor of h (the 1D volume element) on both sides. In 3D you need h^3 instead.

**Reply by Julian-Wyatt** [reply]
I couldn't really improve those errors further with that smoothing range and adjusting the alpha.

When you say multiply both sides of the equation by the volume element, am I correct in saying this is already done in the test case rhs function with the H_x component in:

` barray[i] = numpy.pi ** 2 / 4 * numpy.sin(numpy.pi / 2 * x) * Hx `

and for the left hand side with: 

```
# Equation is scaled by volume element (Hx) for 
# compatibility with the multigrid transfer operators 
# PETSc creates (hence 2/Hx rather than 2/Hx**2), and the
# right hand side is scaled appropriately too.
yarray[i] = 2 / Hx * xarray[iloc]
``` 

So therefore the 1D test function should work considering it is already scaled?

**Comment by annereinarz** [comment]
Sorry, I missed that you were already working with the 1D case, yes the scaling is already introduced there. 

How coarse is your coarsest level and how fine is the finest level?

**Reply by Julian-Wyatt** [reply]
No worries,
The coarsest grid is of size 11 and when the refinements are set to 1,2,3,4 the corresponding loss is (as above) 
[1.73367637e-05 3.07660989e-06 4.58071118e-07 1.54047477e-07]. 
These have 1d finest grid sizes of 21,41,81,161.


However for the 5th refinement ie where the finest grid is 321, the loss is 8.96348653e-01 (as above), which is not of the same order of magnitude, and with a more sensible smooth iterations this is more like: 5.72852890e+00 (alpha = 0.95, pre iters = post iters = 1)

**Reply by annereinarz** [reply]
How are you calculating the norm? The previous points make it unlikely that there's an error in that, but do double-check it. 
How many V-cycle iterations is this taking? Are you sure there's no issues with the coarse solve?

**Reply by Julian-Wyatt** [reply]
It's the same code from the test script to calculate the norm ie
```
    error = expect - u.array_r
    mx, = fine_grid.getSizes()
    Hx = 1 / (mx - 1)
    return Hx * np.linalg.norm(error)
```

For the 5th refinement, it does 100 iterations within mgsolver.solve (as defined with maxiter=100 as it never reaches the tolerance), and I don't think there's any issues with the coarse solve - I've tried the LU decomposition solve function as well as Jacobi with 50 iters and they generally perform similarly, which is why I'm stuck/ confused.

**Reply by annereinarz** [reply]
If you check in your V-cycle I can have a look.

**Reply by Julian-Wyatt** [reply]
I've just pushed it to my repo - thanks for the help!

**Reply by Julian-Wyatt** [reply]
It is worth mentioning - after fixing a silly error in my w-cycle code, that now passes the test with errors of:
`[1.73237317e-05 3.09196226e-06 5.13490925e-07 7.88679903e-08 5.63582524e-09]`, but the V-cycle is still struggling on the final refinement

**Reply by annereinarz** [reply]
I can't see any obvious issues, I would try to use the transpose of the prolongation operator as the restriction instead of using injection. Considering that you only have issues with the finest levels that might be causing the issues.

**Reply by Julian-Wyatt** [reply]
Having made that change and fiddling with the parameters of the solve function to: alpha=0.96 & pre iters = post iters = 2, the test passes with errors:
`[1.73474246e-05 3.06599435e-06 5.42076939e-07 9.58209022e-08 1.84443642e-08]`

Thanks very much! 

Should I consider making this change for W cycle (even though that works)?

**Reply by annereinarz** [reply]
Probably yes, injection is probably the simplest form of restriction, but it does "throw away" a lot of information that could be used.

**Reply by annereinarz** [reply]
Added the restriction issue to the above and marked as answer
